src/yaml-support/schema-holder.ts

```diff
@@ -37,14 +40,49 @@ export class KubernetesClusterSchemaHolder {
 
     private async loadSchemaFromActiveCluster(kubectl: Kubectl, schemaEnumFile?: string): Promise<void> {
         const clusterSwagger = await swagger.getClusterSwagger(kubectl);
-        const crdSchemas = await swagger.getCrdSchemas(kubectl);
-        if (crdSchemas) {
-            this.crdSchemas = crdSchemas;
+        const fetchCRDSchemasEnabled = await this.isFetchingCRDsAllowed(kubectl);
+        if (fetchCRDSchemasEnabled) {
+            const crdSchemas = await swagger.getCrdSchemas(kubectl);
+            if (crdSchemas) {
+                this.crdSchemas = crdSchemas;
+            }
         }
         const schemaRaw = succeeded(clusterSwagger) ? this.definitionsObject(clusterSwagger.result) : util.loadJson(FALLBACK_SCHEMA_FILE());
         this.loadSchemaFromRaw(schemaRaw, schemaEnumFile, kubectl);
     }
 
+    private async isFetchingCRDsAllowed(kubectl: Kubectl): Promise<boolean> {
+        const codeCompletionState = await getSmartCodeCompletionState();
+        if (codeCompletionState !== undefined) {
+            return codeCompletionState === 'enabled';
+        }
+
+        const crdsTypesNumber = await kubectlUtils.getCRDTypesNumber(kubectl);
+        let action: string | undefined;
+        if (crdsTypesNumber > 50) {
+            action = await vscode.window.showInformationMessage(
+                "This cluster contains many CRDs and by fetching them your VSCode instance could slow down. " +
+                "However by disabling CRDs analisys the smart code completion for CR would be automatically disable. " +
+                "Do you want to disable the smart code completion? ",
+                "Disable Once",
+                "Always Disabled",
+                "Enable Once",
+                "Always Enabled"
+                );
+            if (!action
+                || action === 'Disable Once') {
+                return false;
+            } else if (action === 'Always Disabled') {
+                setConfigValue('vs-kubernetes.smart-code-completion', 'disabled');
```

Comment:
Suggest moving this into the `config` module for symmetry with "get state" and so that `config` remains responsible for knowing the key names.